src/Plutarch/Test/QuickCheck/Instances.hs

```diff
@@ -1,864 +1,742 @@
+    cred <- arbitrary
+    scred <- arbitrary
+    pure $
+      Address
+        { addressCredential = cred
+        , addressStakingCredential = scred
+        }
+
+-- | @since 2.1.3
+instance CoArbitrary Address where
+  {-# INLINEABLE coarbitrary #-}
+  coarbitrary addr =
+    coarbitrary (addressCredential addr)
+      . coarbitrary (addressStakingCredential addr)
+
+-- | @since 2.1.3
+instance Function Address where
+  {-# INLINEABLE function #-}
+  function = functionMap into $ \(cred, scred) ->
+    Address
+      { addressCredential = cred
+      , addressStakingCredential = scred
+      }
     where
-      unMap = flip pmatchT $ \(PMap a) -> a
+      into :: Address -> (Credential, Maybe StakingCredential)
+      into addr = (addressCredential addr, addressStakingCredential addr)
 
--- | @since 2.0.0
-instance PArbitrary PPOSIXTime where
-  parbitrary = do
-    (TestableTerm x) <- parbitrary
-    return $ pconT $ PPOSIXTime x
+{- | This generates only those 'TokenName's corresponding to ASCII strings. This
+ is somewhat limited, but otherwise would require UTF-8 encoding as part of
+ the generator. It would also significantly complicate shrinks: we would have
+ to re-encode, shrink, then decode again.
 
-  pshrink = fmap (\(TestableTerm x) -> pconT $ PPOSIXTime x) . shrink . unTime
+ @since 2.1.3
+-}
+instance Arbitrary TokenName where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary =
+    TokenName <$> do
+      ASCIIString name <- arbitrary
+      pure
+        . toBuiltin @ByteString
+        . fromList
+        . fmap (fromIntegral . ord)
+        $ name
+  {-# INLINEABLE shrink #-}
+  shrink (TokenName name) = do
+    let name' =
+          ASCIIString
+            . fmap (chr . fromIntegral)
+            . toList
+            . fromBuiltin
+            $ name
+    ASCIIString name'' <- shrink name'
+    pure
+      . TokenName
+      . toBuiltin @ByteString
+      . fromList
+      . fmap (fromIntegral . ord)
+      $ name''
+
+-- | @since 2.1.3
+deriving via BuiltinByteString instance CoArbitrary TokenName
+
+-- | @since 2.3.1
+instance Function TokenName where
+  {-# INLINEABLE function #-}
+  function = functionMap (coerce @_ @BuiltinByteString) coerce
+
+{- | This generates a phase-1 valid 'Value' meant to represent an /amount/,
+ rather than a /delta/. Thus, all of the following will hold for all generated
+ values and shrinks:
+
+ - Entries whose 'CurrencySymbol' and 'TokenName' match are unique.
+ - \'Outer\' map entries are sorted by 'CurrencySymbol'.
+ - \'Inner\' map entries are sorted by 'TokenName'.
+ - An ADA entry always exists.
+ - All entries have positive amounts.
+
+ While it is possible to have a normalized 'Value' with a zero ADA entry, no
+ such entries are currently generated.
+
+ 'Value's are shrunk only structurally: this means that we can \'drop\'
+ entries, but not shrink existing ones in either their \'inner map\' keys or
+ \'inner map\' values.
+
+ @since 2.1.3
+-}
+instance Arbitrary Value where
```

Comment:
could we add some tests for this? We can grab the tests from PCB for validation; improper normalization in PCB has already ate up a lot of debugging time